File: algorithms/js/validParentheses/validParentheses.js
// Source : https://leetcode-cn.com/problems/valid-parentheses/
// Author : Tian Zhuo
// Date   : 2018-11-18

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    if(s == '') return true;
    if(s.length % 2 == 1 || ')]}'.indexOf(s.charAt(0)) > -1) return false;
    let map = {
        '(': ')',
        '[': ']',
        '{': '}'
    }
    var stack = [s.charAt(0)];
    for (let i = 1; i < s.length; i++) {
        let curr = s.charAt(i);
        if(')]}'.indexOf(curr) > -1 && map[stack[stack.length-1]] == curr) stack.pop();
        else stack.push(curr);
    }
    return stack.length == 0

};

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid1 = function(s) {
    let arr = [];
    for (var char of s) {
        if (char==='(' || char=== '[' || char==='{') {
            arr.unshift(char)
        } else {
            let popStr = arr.shift();
            if (!popStr 
                || popStr==='(' && char!==')' 
                || popStr==='[' && char!== ']' 
                || popStr==='{' && char!=='}') {
                return false;
            }
        }
    }
    return !arr.length;
};

File: src/main/java/validParentheses/ValidParentheses.java
/*
 * @Source: https://leetcode-cn.com/problems/valid-parentheses/
 * @Author: Tian Zhuo
 * @Date: 2018-12-19 20:30:27
 */
package validParentheses;

import java.util.List;
import java.util.ArrayList;

public class ValidParentheses {
    public boolean isValid(String s) {
        List<Character> arr = new ArrayList<>();
        for(int i = 0; i < s.length(); i++) {
            char curr = s.charAt(i);
            if('(' == curr || '[' == curr || '{' == curr){
                arr.add(0, curr);
            }else{
                if(arr.size() == 0) return false;
                char p = arr.get(0);
                arr.remove(0);
                if(p == 0
                    || p == '(' && curr != ')'
                    || p == '[' && curr != ']'
                    || p == '{' && curr != '}'){
                        return false;
                }
            }
        }
        return arr.size() == 0;
    }
}